### PR TITLE
:tada:  New Source: Help Scout

### LINKED ISSUE #2973: New Source: Help Scout
Requirements:
- Build a Help Scout source connector using the [HelpScout Mailbox API](https://developer.helpscout.com/mailbox-api/)
- Build this connector using the [low code CDK](https://docs.airbyte.com/connector-development/config-based/low-code-cdk-overview/)

Streams to support:
- [Conversations](https://developer.helpscout.com/mailbox-api/endpoints/conversations/list/)
- [Conversation Threads](https://developer.helpscout.com/mailbox-api/endpoints/conversations/threads/list/)
- [Customers](https://developer.helpscout.com/mailbox-api/endpoints/customers/list/)
- [Mailboxes](https://developer.helpscout.com/mailbox-api/endpoints/mailboxes/list/)
- [Mailbox Fields](https://developer.helpscout.com/mailbox-api/endpoints/mailboxes/mailbox-fields/)
- [Mailbox Folders](https://developer.helpscout.com/mailbox-api/endpoints/mailboxes/mailbox-folders/)
- [Users](https://developer.helpscout.com/mailbox-api/endpoints/users/list/)
- [Workflows](https://developer.helpscout.com/mailbox-api/endpoints/workflows/list/)
- [Get Satisfaction Rating](https://developer.helpscout.com/mailbox-api/endpoints/ratings/get/)

### PR DESCRIPTION
**Pull request by evwltrs**
## What
A new source for help scout created using the Low Code CDK
Closes #2973

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

**Review by PeerRich — APPROVED** [review]

**Comment by hantmac** [comment]
@evwltrs Hi bro, are u woking on [this pr](https://github.com/airbytehq/airbyte/issues/2973)?

**Comment by evwltrs** [comment]
> @evwltrs Hi bro, are u woking on [this pr](https://github.com/airbytehq/airbyte/issues/2973)?

yep

**Comment by sh4sh** [comment]
Very exciting 🎉 tag me if you have any questions and when you're ready for review!

**Comment by evwltrs** [comment]
> will this include ratings?

yes, we were having issues finding a way to get a rating ID to test the endpoint but we found it so I should be able to finish the connector and push it tonight providing no tech issues

**Comment by evwltrs** [comment]
waiting for Help Scout support to sort out my trial.... they said they had extended it until the 16th but I attempted to log in today and it's expired apprently. 
@misteryeo if you could look into getting Airbyte a permanent testing account that would be useful for the future 👍

**Comment by misteryeo** [comment]
@evwltrs created an issue for the team to get the sandbox: https://github.com/airbytehq/airbyte/issues/22686. @YowanR can help shepherd the creation of the sandbox!

**Comment by evwltrs** [comment]
TODO: Documentation

File: airbyte-integrations/connectors/source-help-scout/integration_tests/invalid_config.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "client_id": "DnGIBWcQvoLekpts3zWMGQasdapbVrd4Ka",
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-help-scout/integration_tests/invalid_config.json:2]: just confirming: are these bogus values that have nothing to do with the real prod values? :)

File: airbyte-integrations/connectors/source-help-scout/source_help_scout/help_scout.yaml [hunk 2]
```diff
@@ -0,0 +1,995 @@
+version: 0.1.0
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-help-scout/source_help_scout/help_scout.yaml:1]: could you rename this file to `manifest.yaml` and update `source.py` to point to the updated file name? we use that filename convention now

File: airbyte-integrations/connectors/source-help-scout/source_help_scout/source.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+source connector.
+
+WARNING: Do not modify this file.
+"""
+
+
+# Declarative Source
+class SourceHelpScout(YamlDeclarativeSource):
+    def __init__(self):
+        super().__init__(**{"path_to_yaml": "help_scout.yaml"})
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-help-scout/source_help_scout/source.py:18]: for when the other file is renamed
```suggestion
        super().__init__(**{"path_to_yaml": "manifest.yaml"})
```

File: airbyte-integrations/connectors/source-help-scout/source_help_scout/schemas/TODO.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
+# TODO: Define your stream schemas
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-help-scout/source_help_scout/schemas/TODO.md:1]: can probably remove this file

File: airbyte-integrations/connectors/source-help-scout/source_help_scout/help_scout.yaml [hunk 5]
```diff
@@ -0,0 +1,995 @@
...
+      client_secret:
+        type: string
+        title: Client secret
+        airbyte_secret: true
+      client_refresh_token:
+        type: string
+        title: Refresh token
+        airbyte_secret: true
+    additionalProperties: true
+  documentation_url: ''
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-help-scout/source_help_scout/help_scout.yaml:994]:
```suggestion
  documentation_url: 'https://docs.airbyte.com'
```

**Review by sherifnada — COMMENTED** [review]
LGTM! 

@brianjlai how should we think about migrating connectors like this one to beta? this one was produced via the UI

**Comment by evwltrs** [comment]
Hi @sherifnada 
Thanks for the review (wading through the giant monolithic connector builder outputs must be a pain). Do you have any idea why I'm failing the spec.yml integration test, I can't recall when I was supposed to generate/make this file.
Thanks

File: airbyte-integrations/connectors/source-help-scout/integration_tests/invalid_config.json (hunk 1 above)
Comment (evwltrs) [reply airbyte-integrations/connectors/source-help-scout/integration_tests/invalid_config.json:2]: Yep, just changed a few digits

**Comment by sherifnada** [comment]
@evwltrs gah, it's an annoyance that we will remove soon. But to unblock you, you just need to put a `spec.yaml` file in the directory where it's expecting it (the one where it can't find it). The file needs to match the one output by your connector when you run the spec command e.g `python main.py spec` 

For context this made sense for python connectors but not so much for declarative connectors. But since this framework is relatively newer not every tool accounts for it yet

**Comment by evwltrs** [comment]
@sherifnada Resolved the issues, let me know if anything else needed. My help scout account expired again so can't pass all the integration tests on my end.

**Comment by danieliser** [comment]
@evwltrs I'd love to test this, have years worth of HelpScout data and a self hosted install on a local Epyc server.

If you can point me in the right direction to installing this and test procedures I'll run it through its paces.

Have full API access though I can't share it currently, would be worth it for me to pay for you to get a temporary account if you can get this done sooner..

feel free to ping me on twitter @daniel_iser or use the email in my github profile to reach out.

**Comment by misteryeo** [comment]
@evwltrs Please make sure to include documentation as part of this PR :)

**Comment by evwltrs** [comment]
Anyone for a final review?

**Comment by misteryeo** [comment]
@evwltrs Did you see my comment about incremental sync support? It looks like several of the endpoints should support this

**Comment by evwltrs** [comment]
> @evwltrs Did you see my comment about incremental sync support? It looks like several of the endpoints should support this

I didn't, no. Where can I find the comment?

**Review by misteryeo — CHANGES_REQUESTED** [review]
Ah I didn't submit the review - here you go @evwltrs

File: docs/integrations/sources/help-scout.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+4. Click **Set up source**.
+
+## Supported sync modes
+
+The Help Scout source connector supports the following [sync modes](https://docs.airbyte.com/cloud/core-concepts#connection-sync-modes):
+
+| Feature                       | Supported? |
+| :---------------------------- | :--------- |
+| Full Refresh Sync             | Yes        |
+| Incremental Sync              | No         |
```
[the remark this answers is not in the document]
Comment (evwltrs) [reply docs/integrations/sources/help-scout.md:46]: Hi Andy, thanks for the review and apologies for some of the slip ups, I'll fix them when I get home. I have been skimming through the docs and I haven't managed to find any info on how to understand what streams can be incremental syncs and what streams cannot. If I missed this part could you please link it 🙏

Thanks

[the remark this answers is not in the document]
Comment (evwltrs) [reply docs/integrations/sources/help-scout.md:46]: @misteryeo ^

[the remark this answers is not in the document]
Comment (evwltrs) [reply docs/integrations/sources/help-scout.md:46]: @misteryeo might have got buried in your inbox

[the remark this answers is not in the document]
Comment (misteryeo) [reply docs/integrations/sources/help-scout.md:46]: @evwltrs apologies for the delay here. More information can be found in the docs here: https://docs.airbyte.com/connector-development/config-based/understanding-the-yaml-file/incremental-syncs/ but generally anything that has some sort of cursor value like a timestamp should support incremental syncs!

[the remark this answers is not in the document]
Comment (evwltrs) [reply docs/integrations/sources/help-scout.md:46]: @sherifnada Could I also be invited to the Help Scout account on Lastpass please 🙏

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

**Comment by evwltrs** [comment]
@sherifnada Could I also be invited to the Help Scout account on Lastpass please 🙏

**Comment by sherifnada** [comment]
@YowanR could you help on that?

**Comment by evwltrs** [comment]
> @YowanR could you help on that?

still waiting on this 😬

**Comment by evwltrs** [comment]
Anyone?

**Comment by marcosmarxm** [comment]
@evwltrs sorry the long delay here. @sh4sh can you reach out Evans and help finishing this contributor-program contribution?

**Comment by evwltrs** [comment]
Ok, back on working on this today, apologies for the really long delay, hopefully everything is still valid. Reducing the amount of user settings and adding incremental sync.

**Comment by jrolom** [comment]
hi @evwltrs  - are you still planning to work on this?

**Comment by evwltrs** [comment]
hi @jrolom 
yes! I am definitely still working on it. I just completely forgot :pensive: 
I am going to try and figure out incremental syncs tonight.

Thanks

**Comment by evwltrs** [comment]
Is the bounty still valid or have I left it too long?

**Comment by jrolom** [comment]
Hi @evwltrs ! yeah, it's still valid, it would be unfair if we changed it half way through your contribution.

**Comment by jrolom** [comment]
Hey @evwltrs ! Will you be working on this?

**Comment by marcosmarxm** [comment]
I'm going to close due inactivity. @evwltrs please let us know if you want to continue this work so I can reopen it.

**Comment by catkins** [comment]
Hi there, I just realised that @evwltrs had already done this _after_ I'd opened a ticket noting that I'd created a custom connector for our OSS instance.

I dumped out the yaml

**Comment by catkins** [comment]
Hi there, I just realised that @evwltrs had already done this _after_ I'd opened a ticket noting that I'd created a custom connector for our OSS instance.

I dumped out the yaml into this gist, but didn't really spend much time on tidying the schema, so its just using the inferred schema which is a bit message (with _links and stuff embedded)

https://gist.github.com/catkins/d1267c32fda475d54ce5a91b6015dcd3

I did manage to get incremental loading working for conversations btw. We can perhaps join forces to get this PR revived and in the right shape?

File: docs/integrations/sources/help-scout.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
+# Help Scout
+
+This page contains the setup guide and reference information for the Help Scout source connector.
+
+## Prerequisites
+
+1. Create a Help Scout account.
+2. Create an application and note the App ID and App Secret.
+3. Navigate to `https://secure.helpscout.net/authentication/authorizeClientApplication?client_id={application_id}&state={your_secret}`
```
Comment (catkins) [inline docs/integrations/sources/help-scout.md:9]: This isn't required - if you use the client-credentials flow, you can just use the app id as the client id, and app secret as the client secret in Airbyte

https://developer.helpscout.com/mailbox-api/overview/authentication/#client-credentials-flow

**Comment by evwltrs** [comment]
@catkins This sounds great. I was very close but couldn't figure out the incremental syncs. Let me know if you need anything.

**Comment by catkins** [comment]
Rad. I'll see if I can branch off your PR, rebase it and incorporate my changes.

**Comment by kelechi2020** [comment]
@evwltrs @catkins  Incremental sync in the helpscout connector may be super tricky using Low code implementation due to the way Helpscout API implements filtering records by date.
https://developer.helpscout.com/mailbox-api/endpoints/conversations/list/#query 

Has anyone tried to find a way around this ?

**Comment by catkins** [comment]
So you can pass `modifiedSince` as a query param, but annoyingly the conversation type itself doesn't include a modified at property. 
What I ended up doing was using created at as the cursor field and just over fetching a bit on each daily sync, and deduped in dbt.

**Comment by kelechi2020** [comment]
Thanks @catkins . I used `modifiedSince`and createdAt in the format you mentioned and the behavior was not what I expected.
How did you handle deduping with dbt ? You can point me to a doc or something.

**Comment by kelechi2020** [comment]
Has anyone tried building a custom connector using airbyte python cdk to use Helpscout's query interface https://developer.helpscout.com/mailbox-api/endpoints/conversations/list/#query ?

**Comment by catkins** [comment]
https://github.com/dbt-labs/dbt-utils#deduplicate-source was what I used to deduplicate. 

I'll export my YAML for the connector when I'm back at the computer on Monday

**Comment by valentinfily** [comment]
Any plan to re-open this to support Helpscout in Airbyte ?

**Comment by tylerdigital** [comment]
I'm considering building out an incremental source for HelpScout, I just wanted to check if any effort is underway currently? If there's an official connector, I'd definitely prefer to use that than building/maintaining on our own

**Comment by catkins** [comment]
Apologies, priorities shifted at work and I hadn't looped back to it

**Comment by booleanbetrayal** [comment]
Would also be interested in a Help Scout connector and could likely assist in testing.

**Comment by valentinfily** [comment]
Can we please reopen this @marcosmarxm ?

**Pull request closed without merging**